core/src/main/java/org/apache/iceberg/BaseRewriteManifests.java

```diff
@@ -295,7 +305,10 @@ private WriterWrapper getWriter(Object key) {
   @Override
   protected void cleanUncommitted(Set<ManifestFile> committed) {
     cleanUncommitted(newManifests, committed);
-    cleanUncommitted(addedManifests, committed);
+    if (!snapshotIdInheritanceEnabled) {
```

Comment:
Is this correct? I thought that added manifests are owned by the table only if inheritance is allowed. That would mean that added manifests are only removed if inheritance is allowed, right?

Reply:
If inheritance is enabled, `addedManifests` contains original manifests that should not be removed no matter what the operation outcome is. If the commit fails, the caller can retry. If the commit succeeds, the manifests are part of the metadata now. 

If inheritance is not enabled, `addedManifests` will contain manifest copies. Those must be always cleaned up as the caller doesn't have access to them.

I tried to summarize that in the description to `AppendFiles`.

> By default, the manifest will be rewritten to assign all entries this update's snapshot ID.
> In that case, it is always the responsibility of the caller to manage the lifecycle of
> the original manifest.
>
> If manifest entries are allowed to inherit the snapshot ID assigned on commit, the manifest
> should never be deleted manually if the commit succeeds as it will become part of the table
> metadata and will be cleaned up on expiry. If the manifest gets merged with others while
> preparing a new snapshot, it will be deleted automatically if this operation is successful.
> If the commit fails, the manifest will never be deleted and it is up to the caller whether
> to delete or reuse it.

Reply:
Like in the other cleanup method, should this check whether committed is empty?

Reply:
Okay, so I think the difference between this and the logic in `MergingSnapshotProducer` is that this won't compact and rewrite the committed manifests. That means that if the commit succeeded, then all the added manifests are part of the table (so we know they will all be in the committed list).

Here's a quick summary:

* Inheritance is enabled
  * Commit succeeded - do not delete, the files are owned and part of the table
  * Commit failed - do not delete, the files are not owned
* Inheritance is not enabled - added manifests are rewritten and are owned
  * Commit succeeded - run normal manifest cleanup (rely on the committed set)
  * Commit failed - run normal manifest cleanup (committed set will be empty)

Reply:
Yes, the summary is correct. In `MergingSnapshotProducer`, though, we can merge the appended manifest while preparing a new snapshot and it will never be part of the table metadata.